### PR TITLE
Add support for ticket drop times

### LINKED ISSUE #655: Add a `ticket_drop_time` to control ticket visibility and editing
`create_tickets` currently deletes and recreates tickets / other parameters around the event every time it is called

This could be very problematic if a bought ticket is deleted since the organizer decides to go in and add a few more tickets.

We should:
- Add a parameter to events specifying a `ticket_drop_time`
- Make tickets publicly available for purchase only after this time
- Freeze ticket offering editing after this time

### PR DESCRIPTION
**Pull request by aviupadhyayula**
Changelog: 
1. Allows admins to specify when tickets will "drop" in `create_tickets`. After this "drop time" has passed, ticket offerings cannot be edited. 
2. Tickets cannot be added to cart until the drop time has elapsed. 
3. Tickets cannot be edited after they've been sold or checked out in a user's cart.

Closes #655.

### CODE REVIEW

File: backend/Pipfile [hunk 1]
```diff
@@ -56,6 +56,7 @@ numpy = "*"
 inflection = "*"
 cybersource-rest-client-python = "*"
 pyjwt = "*"
+freezegun = "*"
```
Comment (aviupadhyayula) [inline backend/Pipfile:59]: Used in the unit tests to simulate 12 hours passing after an event's creation. [Here's the library's repo](https://github.com/spulec/freezegun).

Comment (rohangpta) [reply backend/Pipfile:59]: FYI, you might have used `pipenv install` which updates packages where possible. For example, ruff is now pinned to 0.4.1 in the lock file (which may be breaking with previous versions). This is probably not your intended behavior, so you might want to revert and do a `pipenv upgrade` instead 

[Reading](https://stackoverflow.com/questions/55592586/how-to-install-update-package-with-pipenv-without-updating-the-rest-of-package)

File: backend/clubs/views.py [hunk 2]
```diff
@@ -2707,6 +2724,11 @@ def create_tickets(self, request, *args, **kwargs):
             event.ticket_order_limit = order_limit
             event.save()
 
+        delay_drop = request.data.get("delay_drop", None)
```
Comment (rohangpta) [inline backend/clubs/views.py:2727]: Why + 12h here? Seems arbitrary. Can we maybe give the user better control on when exactly their tickets drop?

File: backend/tests/clubs/test_ticketing.py [hunk 3]
```diff
@@ -155,6 +156,43 @@ def test_create_ticket_offerings_bad_data(self):
...
+        # Drop time should be set
+        self.assertIsNotNone(self.event1.ticket_drop_time)
+
+        # Drop time should be 12 hours from initial ticket creation
+        expected_drop_time = timezone.now() + timezone.timedelta(hours=12)
+        diff = abs(self.event1.ticket_drop_time - expected_drop_time)
+        self.assertTrue(diff < timezone.timedelta(minutes=5))
+
+        # Move Django's internal clock 13 hours forward
+        with freeze_time(timezone.now() + timezone.timedelta(hours=13)):
```
Comment (rohangpta) [inline backend/tests/clubs/test_ticketing.py:186]: Nice usage of an external lib!

**Review by rohangpta — CHANGES_REQUESTED** [review]

File: backend/clubs/views.py (hunk 2 above)
Comment (aviupadhyayula) [reply backend/clubs/views.py:2727]: No particular reason behind choosing 12. Will modify `delay_drop` to accept an integer value, and use that instead.

Comment (rohangpta) [reply backend/clubs/views.py:2727]: Why not just a timestamp?

Comment (aviupadhyayula) [reply backend/clubs/views.py:2727]: Like a Unix timestamp? Sure.

Comment (rohangpta) [reply backend/clubs/views.py:2727]: Not necessarily a Unix timestamp...there's some use cases around the codebase where we pass in datetime from frontend to backend. For example, we allow club admins to control the `application_start_time`, `application_end_time`, etc. You could use those as guiding examples perhaps

File: backend/Pipfile (hunk 1 above)
Comment (aviupadhyayula) [reply backend/Pipfile:59]: Thanks for the link, was helpful. Fixed.

**Comment by aviupadhyayula** [comment]
Added new feats in later commits (i.e. admins cannot edit tickets after they've been sold or checked out). Re-requesting a review.

**Comment by aviupadhyayula** [comment]
Thinking about this more: it's weird if users see tickets as available in `/event/tickets`, but can't add them to their cart. Thoughts on only displaying tickets as available once the drop time has elapsed?

File: backend/clubs/views.py [hunk 4]
```diff
@@ -2618,6 +2625,10 @@ def create_tickets(self, request, *args, **kwargs):
                             order_limit:
                                 type: int
                                 required: false
+                            drop_time:
+                                type: string
+                                format: date-time
```
Comment (rm03) [inline backend/clubs/views.py:2630]: nit: can you update the response schema to reflect 403

**Comment by rm03** [comment]
> Thinking about this more: it's weird if users see tickets as available in `/event/tickets`, but can't add them to their cart. Thoughts on only displaying tickets as available once the drop time has elapsed?

Agreed, this would be better

File: backend/clubs/views.py (hunk 4 above)
Comment (aviupadhyayula) [reply backend/clubs/views.py:2630]: Thanks for catching this, @rm03. Your commitment to detail is bot-like. Appreciate your kind review.

**Comment by rohangpta** [comment]
Nice work! Afterthought: would it be cool to show a countdown to ticket drop time on the frontend?

**Comment by aviupadhyayula** [comment]
> Nice work! Afterthought: would it be cool to show a countdown to ticket drop time on the frontend?

Yeah, that's a good idea. cc @julianweng would be cool to put this in.

**Pull request merged**